src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs

```diff
@@ -919,6 +942,110 @@ internal override void TryForceResolve(bool asValueType)
             }
         }
 
+        /// <summary>
+        /// Extension for a type argument whose nullable annotation after substitution depends on the type constraints on the type parameter.
+        /// To avoid a cycle, we delay determining the nullable annotation until after the type parameter's type constraints are bound.
+        /// </summary>
+        private sealed class LazySubstitutedType : Extensions
+        {
+            private readonly ImmutableArray<CustomModifier> _customModifiers;
+            private readonly TypeParameterSymbol _typeParameter;
+
+            private const int Unresolved = -1;
+            private int _resolved;
+
+            public LazySubstitutedType(ImmutableArray<CustomModifier> customModifiers, TypeParameterSymbol typeParameter)
+            {
+                Debug.Assert(!customModifiers.IsDefault);
+                _customModifiers = customModifiers;
+                _typeParameter = typeParameter;
+                _resolved = Unresolved;
+            }
+
+            internal override SpecialType GetSpecialType(TypeSymbol typeSymbol) => typeSymbol.SpecialType;
+            internal override bool IsRestrictedType(TypeSymbol typeSymbol, bool ignoreSpanLikeTypes) => typeSymbol.IsRestrictedType(ignoreSpanLikeTypes);
+            internal override bool IsStatic(TypeSymbol typeSymbol) => typeSymbol.IsStatic;
+            internal override bool IsVoid(TypeSymbol typeSymbol) => typeSymbol.IsVoidType();
+            internal override bool IsSZArray(TypeSymbol typeSymbol) => typeSymbol.IsSZArray();
+            internal override bool IsRefLikeType(TypeSymbol typeSymbol) => typeSymbol.IsRefLikeType;
+
+            internal override NullableAnnotation GetResolvedAnnotation(NullableAnnotation defaultAnnotation)
+            {
+                Debug.Assert(defaultAnnotation == NullableAnnotation.Ignored);
+                if (_resolved == Unresolved)
+                {
+                    Interlocked.CompareExchange(ref _resolved, value: (int)getResolvedAnnotationCore(), comparand: Unresolved);
+                }
+
+                Debug.Assert(_resolved != Unresolved);
+                return (NullableAnnotation)_resolved;
+
+                NullableAnnotation getResolvedAnnotationCore()
```

Comment:
static?

Reply:
I feel satisfied with the current factoring (capturing 'this'). Thanks

Reply:
Oh, I'd missed that. Thanks